synapse/storage/databases/main/devices.py

```diff
@@ -573,16 +576,53 @@ async def _get_cached_user_device(self, user_id: str, device_id: str) -> JsonDic
         return db_to_json(content)
 
     @cached()
-    async def get_cached_devices_for_user(self, user_id: str) -> Dict[str, JsonDict]:
-        devices = await self.db_pool.simple_select_list(
-            table="device_lists_remote_cache",
-            keyvalues={"user_id": user_id},
-            retcols=("device_id", "content"),
-            desc="get_cached_devices_for_user",
+    async def get_cached_devices_for_user(
+        self, user_id: str
+    ) -> Optional[Dict[str, JsonDict]]:
+        """Retrieve the most recent cached devices data.
+
+        We can be in three states, depending on the latest response from the remote
+        homeserver.
+
+        - We could never requested devices for this user. In this case, return `None`.
+        - We could have requested devices for this user, only to be told they don't
+          have any devices. In this case, return an empty dictionary.
+        - Otherwise, we've cached details of 1 or more devices for this user. Return
+          a a dictionary from device id to the device data.
+        """
+        return await self.db_pool.runInteraction(
+            "get_cached_devices_for_user",
+            self._get_cached_devices_for_user_txn,
+            user_id,
         )
-        return {
-            device["device_id"]: db_to_json(device["content"]) for device in devices
-        }
+
+    def _get_cached_devices_for_user_txn(
+        self, txn: LoggingTransaction, user_id: str
+    ) -> Optional[Dict[str, JsonDict]]:
+        # Four cases:
+        # 1. No stream id, no cached devices. Query yields no rows. Return None.
+        # 2. No stream id, >= 1 cached devices. Invalid state. Query will yield no rows.
+        #    return None.
+        # 3. Stream id, no cached devices. Return empty dict. Query returns one row
+        #    (non-NULL stream_id, NULL, NULL). Return empty dict.
+        # 4. Stream id, >= 1 cached devices. Query return 1 or more row
+        #    (non-NULL stream_id, non-NULL device_id, non-NULL content). Return dict.
+        query = """
+            SELECT stream_id, device_id, content
+            FROM device_lists_remote_extremeties
+                LEFT JOIN device_lists_remote_cache USING(user_id)
+            WHERE user_id = ?
+        """
+        txn.execute(query, (user_id,))
+        devices: List[Tuple[str, Optional[str], Optional[str]]] = txn.fetchall()
+        if not devices:
+            return None
+        elif devices[0][1] is None:
+            return {}
```

Comment:
Should we check the length here?

Reply:
If `devices` is truthy (618) it has at least one entry. But having said that, FWIW I think this could be clearer if I made two separate queries: one to see if there's an entry in the extremities table, then another to read from the cache table. 

I wanted to write a single query here---in the spirit of the postgres book----as an exercise and as a bit of practice to see how it came out. I'm not really convinced by the result TBH. Maybe two queries is better. It's not like this is going to be an application bottleneck anyway.

Reply:
Right there will be at least one query, but what is it possible to get multiple rows and have one of them (the first one) but `<something>, NULL, NULL`? Or is that impossible from the query?

Reply:
The changes to the query here have been reverted.